### DISCUSSION TITLE
Migrating from 1.13 to 2 (issue with chartModel) [NEWBIE]

### DISCUSSION THREAD

**Discussion by Prince1082** [patrykandpatrick/vico#565] [Questions] [answered]
Hello everyone,

I'm currently working on a project where I fetch data from an API and aim to display it in a graph. Initially, I made the API graph work in version 1.13, but there was no axis override, and the rate data was high with less fluctuation, resulting in the graph looking like an almost straight line. To address this, I switched to vico2.

I have created a sample list for testing purposes, but unfortunately, the graphs are not being rendered correctly. I have implemented logging to track the data, and it appears to be correct. However, I'm struggling to identify the issue in my implementation. (It seems like the mapping might be the issue, but I don't know how to fix it.) Additionally, if there is any better way to do it, please let me know.

If anyone could provide guidance or assistance in pinpointing the problem, it would be greatly appreciated. Below is a snippet of my code for reference:

```
data class Rate(
    val date: String,
    val silRate: Int,
    val gbRate: Int,
    val ghRate: Int,
    val gmRate: Int,
)

val sampleRates = listOf(
    Rate("01-01-2024", 720, 54632, 35000, 46354),
    Rate("02-01-2024", 750, 55432, 42000, 47543),
    Rate("03-01-2024", 730, 56543, 54200, 48456),
    Rate("04-01-2024", 740, 54321, 22050, 45432),
    Rate("05-01-2024", 720, 57654, 65400, 49543),
    Rate("06-01-2024", 710, 55432, 45400, 46543)
)
val simpleDateTimeFormatter: DateTimeFormatter = DateTimeFormatter.ofPattern("dd-MM-yyyy")

@Composable
fun LineChartScreen() {
    val formatted = sampleRates.map {
        val date = LocalDate.parse(it.date, simpleDateTimeFormatter)
        val datedm = date.toEpochDay()
        Log.d("ChartingApp", "Processing Rate: $date ($datedm)")
        datedm to it.ghRate
    }.onEach { (key, value) ->
        Log.d("ChartingApp", "Mapped key: $key, value: $value")
    }

    val dateTimeFormatter = DateTimeFormatter.ofPattern("d MMM")
    val horizontalAxisValueFormatter =
        AxisValueFormatter<AxisPosition.Horizontal.Bottom> { value, _, _ ->
            LocalDate.ofEpochDay(value.toLong()).format(dateTimeFormatter)
        }

    Column(Modifier.fillMaxSize()) {
        Text(text = "Rates")
        CartesianChartHost(
            chart = rememberCartesianChart(
                rememberLineCartesianLayer(),
                startAxis = rememberStartAxis(),
                bottomAxis = rememberBottomAxis(valueFormatter = horizontalAxisValueFormatter)
            ),

            model = CartesianChartModel(
                LineCartesianLayerModel.build {
                formatted.forEach { (datedm, ghRate) ->
                    Log.d("ChartingApp", "Adding series: key: $datedm, value: $ghRate")
                    series(datedm, ghRate)
                }
            }),
        )
    }
}
```
Sorry for the bad ScreenShots
Here the graph is showing 6 different line while i want it to show 1 line for 6 different dates:
![Screenshot 2024-02-04 015604](https://github.com/patrykandpatrick/vico/assets/56156635/2f8d192a-8205-4cbc-a139-6e7f8eba81a3)
logged data:
![Screenshot 2024-02-04 015644](https://github.com/patrykandpatrick/vico/assets/56156635/71080aa8-d54b-4f51-bfb4-e87f13fce6d7)

**Comment by patrickmichalik** [comment, marked as answer]
Hello! One line corresponds to one series, but you seem to be adding six series. Please try replacing `map` with `associate` in `sampleRates.map` and using the following:

```kt
LineCartesianLayerModel.build { series(formatted.keys, formatted.values) }
```

Please also note that mapping data on the main thread on each recomposition can significantly degrade performance. These operations should happen off the main thread, preferably in a `ViewModel`. Another option is to use `LaunchedEffect` and `withContext(Dispatchers.Default)`, for example. So long as `sampleRates` remains unchanged, no mapping should occur, and the same `CartesianChartModel` should be reused. I’d also suggest using `remember` for `dateTimeFormatter` and `horizontalAxisValueFormatter` or moving them out of the composable function.

**Reply by Prince1082** [reply]
Hello! Thank you for your prompt response.

I appreciate your insight. I've taken your suggestion and replaced map with associate in sampleRates.map and it worked, also I tried it with the actual api data and it worked, as of now it uses data from the shared preferences where I keep api data as cache.

Regarding your performance tips, I'll make the necessary adjustments. Moving data mapping operations off the main thread, possibly into a ViewModel or using LaunchedEffect with withContext(Dispatchers.Default)

Thanks again for your help! I'll implement these changes and see how it impacts the graph rendering.

**Reply by patrickmichalik** [reply]
You’re welcome! I’m glad to hear that this worked.
